lib/fswatcher/fswatcher.go

```diff
@@ -0,0 +1,496 @@
+	// children.
+	localMaxFilesPerDir := maxFiles
+	var currPath string
+	for i, pathSegment := range pathSegments[:len(pathSegments)-1] {
+		currPath = filepath.Join(currPath, pathSegment)
+
+		if event, ok := parentDir.events[currPath]; ok {
+			event.lastModTime = eventTime
+			event.eventType |= eventType
+			l.Debugf("%v Parent %s (type %s) already tracked: %s",
+				watcher, currPath, event.eventType, path)
+			return
+		}
+
+		if parentDir.childCount() == localMaxFilesPerDir {
+			l.Debugf("%v Parent dir %s already has %d children,"+
+				"tracking it instead: %s", watcher, currPath,
+				localMaxFilesPerDir, path)
+			watcher.aggregateEvent(filepath.Dir(currPath),
+				eventTime, eventType)
+			return
+		}
+
+		// If there are no events below path, but we need to recurse
+		// into that path, create eventDir at path.
+		if _, ok := parentDir.dirs[currPath]; !ok {
+			l.Debugf("%v Creating eventDir: %s", watcher, currPath)
+			parentDir.dirs[currPath] = newEventDir(currPath,
+				parentDir)
+		}
+		parentDir = parentDir.dirs[currPath]
+
+		// Reset allowed children count to maxFilesPerDir for non-root
+		if i == 0 {
+			localMaxFilesPerDir = maxFilesPerDir
+		}
+	}
+
+	if event, ok := parentDir.events[path]; ok {
+		event.lastModTime = eventTime
+		event.eventType |= eventType
+		l.Debugf("%v Already tracked (type %v): %s", watcher,
+			event.eventType, path)
+		return
+	}
+
+	childDir, ok := parentDir.dirs[path]
+
+	// If a dir existed at path, it would be removed from dirs, thus
+	// childCount would not increase.
+	if !ok && parentDir.childCount() == localMaxFilesPerDir {
+		l.Debugf("%v Parent dir already has %d children, tracking it "+
+			"instead: %s", watcher, localMaxFilesPerDir, path)
+		watcher.aggregateEvent(filepath.Dir(path), eventTime, eventType)
+		return
+	}
+
+	firstModTime := eventTime
+	if ok {
+		firstModTime = childDir.getFirstModTime()
+		eventType |= childDir.getEventType()
+		delete(parentDir.dirs, path)
+	}
+	l.Debugf("%v Tracking (type %v): %s", watcher, eventType, path)
+	parentDir.events[path] = &fsEvent{path, firstModTime, eventTime, eventType}
+	watcher.resetNotifyTimerIfNeeded()
+}
+
+func (watcher *fsWatcher) actOnTimer() {
+	eventCount := watcher.rootEventDir.eventCount()
+	if eventCount == 0 {
+		l.Debugln(watcher, "No tracked events, waiting for new event.")
+		watcher.notifyTimerNeedsReset = true
+		return
+	}
+	oldFsEvents := watcher.popOldEvents(watcher.rootEventDir, time.Now())
+	// Sending to channel might block for a long time, but we need to keep
+	// reading from notify backend channel to avoid overflow
+	if len(oldFsEvents) != 0 {
+		go func() {
+			timeBeforeSending := time.Now()
+			l.Debugln("%v Notifying about %d fs events", watcher,
+				len(oldFsEvents))
+			separatedBatches := make(map[fsEventType][]string)
+			separatedBatches[nonRemove] = []string{}
+			separatedBatches[mixed] = []string{}
+			separatedBatches[remove] = []string{}
```

Comment:
You don't need this

x := append(nil, "blah") results in x := []string{"blah"}